### PR TITLE
[MAINTENANCE] Refactoring Profile Code 

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* Introduced typed `Parameter` object
* Code refactored such that modules fit configuration entities for consistency and a better code readability
* Type Hints for a more rapid code traversal and understanding
* All of tests created by @anthonyburdi pass (modified to fit the new structure)
* Upstream `develop` was also merged in to this working branch

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [ENHANCEMENT], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
-
-
-

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

Thank you for submitting!

### CODE REVIEW

File: great_expectations/profiler/configuration_builder/parameter_id_configuration_builder.py [hunk 1]
```diff
@@ -1,7 +1,8 @@
...
-
-from ..rule_state import RuleState
-from .configuration_builder import ConfigurationBuilder
+from great_expectations.profiler.configuration_builder.configuration_builder import (
+    ConfigurationBuilder,
+)
+from great_expectations.profiler.profiler_rule.rule_state import RuleState
 
 
 class ParameterIdConfigurationBuilder(ConfigurationBuilder):
```
Comment (talagluck) [inline great_expectations/profiler/configuration_builder/parameter_id_configuration_builder.py]: I'm wondering if we can/should change the name of this class for clarity.

File: great_expectations/profiler/domain_builder/column_domain_builder.py [hunk 2]
```diff
@@ -1,25 +1,29 @@
...
-from .domain_builder import DomainBuilder
+import great_expectations.exceptions as ge_exceptions
+from great_expectations.execution_engine.execution_engine import MetricDomainTypes
+from great_expectations.profiler.domain_builder.domain_builder import DomainBuilder
+from great_expectations.validator.validator import Validator
 
 
 # TODO: <Alex>ALEX -- The fact that this class overwrites get_domains() instead of _get_domains() is problematic and should be changed to the standarized approach (the latter).</Alex>
 class ColumnDomainBuilder(DomainBuilder):
+    # TODO: <Alex>ALEX -- What is the return type?</Alex>
```
Comment (talagluck) [inline great_expectations/profiler/domain_builder/column_domain_builder.py]: I would think this returns a dictionary (or a list of dictionaries) of domain kwargs

File: great_expectations/profiler/domain_builder/column_domain_builder.py [hunk 3]
```diff
@@ -1,25 +1,29 @@
...
 
-from ..exceptions import ProfilerConfigurationError
-from .domain_builder import DomainBuilder
+import great_expectations.exceptions as ge_exceptions
+from great_expectations.execution_engine.execution_engine import MetricDomainTypes
+from great_expectations.profiler.domain_builder.domain_builder import DomainBuilder
+from great_expectations.validator.validator import Validator
 
 
 # TODO: <Alex>ALEX -- The fact that this class overwrites get_domains() instead of _get_domains() is problematic and should be changed to the standarized approach (the latter).</Alex>
```
Comment (talagluck) [inline great_expectations/profiler/domain_builder/column_domain_builder.py]: Agreed!

File: great_expectations/profiler/domain_builder/simple_semantic_type_domain_builder.py [hunk 4]
```diff
@@ -55,13 +56,17 @@ def _get_domains(
...
             raise ValueError("unrecognized")
-        columns = validator.get_metric(MetricConfiguration("table.columns", dict()))
-        domains = []
+        domains: List[Dict[str, Union[str, MetricDomainTypes, Dict[str, Any]]]] = []
+        columns: List[str] = validator.get_metric(
+            MetricConfiguration("table.columns", {})
+        )
+        column: str
         for column in columns:
             # TODO: <Alex>ALEX -- Is this "column_type"?  Or is this more of a "domain type" (or "the name of the domain type")?</Alex>
```
Comment (talagluck) [inline great_expectations/profiler/domain_builder/simple_semantic_type_domain_builder.py]: I believe this is a semantic type, which is distinguished from the column storage type. A storage type would be something like `int`. The semantic type would be something like `id`.

File: great_expectations/profiler/parameter_builder/multi_batch_parameter_builder.py [hunk 5]
```diff
@@ -1,25 +1,28 @@
...
+
     MultiBatchParameterBuilder checks that there are multiple batch ids passed to build_parameters,
     and uses a configured batch_request parameter to obtain them if they are not.
     """
 
     def __init__(self, *, parameter_id, batch_request, data_context):
         if data_context is None:
-            raise ProfilerExecutionError(
-                f"MultiBatchParameterBuilder requires a data_context, but none was provided."
+            raise ge_exceptions.ProfilerExecutionError(
```
Comment (talagluck) [inline great_expectations/profiler/parameter_builder/multi_batch_parameter_builder.py:32]: Why the change from `ProfilerExecutionError` to `ge_exceptions.ProfilerExecutionError`?

File: great_expectations/profiler/parameter_builder/parameter.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+from dataclasses import asdict, dataclass
+from typing import Any, Dict, Optional, Union
+
+from great_expectations.core import IDDict
+from great_expectations.core.util import convert_to_json_serializable
+from great_expectations.types import SerializableDictDot
+
+
+@dataclass
```
Comment (talagluck) [inline great_expectations/profiler/parameter_builder/parameter.py:8]: Lovely!

File: great_expectations/profiler/parameter_builder/parameter.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+class Parameter(IDDict, SerializableDictDot):
```
Comment (talagluck) [inline great_expectations/profiler/parameter_builder/parameter.py]: As discussed, I think it makes sense to start to do away with `DictDot` and `SerializableDictDot`, since we can accomplish the same functionality with `dataclasses`. Maybe recreating `SerializableDictDot` as `SerializableDataclass`.

File: great_expectations/profiler/profiler_rule/profiler_rule.py [hunk 8]
```diff
@@ -1,19 +1,27 @@
...
 class ProfilerRule:
     def __init__(
         self,
         name: str,
         domain_builder: DomainBuilder,
-        parameter_builders: [ParameterBuilder],
-        configuration_builders: [ConfigurationBuilder],
+        parameter_builders: List[ParameterBuilder],
+        configuration_builders: List[ConfigurationBuilder],
+        # TODO: <Alex>ALEX -- what should the type of "variables" be?</Alex>
```
Comment (talagluck) [inline great_expectations/profiler/profiler_rule/profiler_rule.py]: Based on your refactor of variables into a typed Parameter object, I would expect that variables would be a dictionary of Parameters.

File: great_expectations/profiler/profiler_rule/rule_state.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+from great_expectations.profiler.parameter_builder.parameter import Parameter
+
+DOMAIN_KWARGS_PARAMETER_NAME: str = "domain_kwargs"
+DOMAIN_KWARGS_PARAMETER_FULLY_QUALIFIED_NAME: str = (
+    f"$domain.{DOMAIN_KWARGS_PARAMETER_NAME}"
+)
+VARIABLES_KEY: str = "$variables."
+
+
+# TODO: <Alex>ALEX -- We should consider changing the names "parameters" and "Parameter.parameters" to be more semantically meaningful.</Alex>
```
Comment (talagluck) [inline great_expectations/profiler/profiler_rule/rule_state.py:15]: 💯

File: great_expectations/profiler/profiler_rule/rule_state.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        """
+        Getter for the id of the rule domain
+        :return: the id of the rule domain
+        """
+        return IDDict(self.active_domain).to_id()
+
+    def get_parameter_value(self, parameter_name: str) -> Any:
+        """
+        Get a value from the current rule state. Values must be dot-delimited, and may start either with
+        the key 'domain' or the id of a parameter.
```
Comment (talagluck) [inline great_expectations/profiler/profiler_rule/rule_state.py]: Is the id of a parameter meaningful? I thought that was the unique identifier that we had looked at? Or does this mean more like `$mean`? Are `$mean` and that long unique identifier both parameter IDs?

File: tests/profiler/test_rule_state.py [hunk 11]
```diff
@@ -20,27 +21,33 @@ def semantic_rule_state():
     return RuleState(
         active_domain={"domain_kwargs": {"column": "Age"}, "semantic_type": "numeric"},
         domains=[{"domain_kwargs": {"column": "Age"}, "semantic_type": "numeric"}],
-        parameters={"f45a40fda1738351c5e67a0aa89c2c7c": {"mean": 5.0}},
-        variables={"false_positive_threshold": 0.01},
+        parameters={
+            "f45a40fda1738351c5e67a0aa89c2c7c": Parameter(
```
Comment (talagluck) [inline tests/profiler/test_rule_state.py:25]: This `f45...` is what I was describing above. Is this the `parameter_id` or is `mean` the `parameter_id`?  Or is this the `domain_id` (if that is the case, I think this solves some of our problems, though I would then expect that there should be multiple parameters allowed per domain, and so this should be a list or dictionary).

**Review by talagluck — COMMENTED** [review]
Looks great overall! A bunch of questions and comments. Leaving unapproved for now, and will take another look at your responses with fresh eyes tomorrow.

File: great_expectations/profiler/configuration_builder/parameter_id_configuration_builder.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/configuration_builder/parameter_id_configuration_builder.py]: @talagluck Good idea -- will try in next commit.

File: great_expectations/profiler/domain_builder/column_domain_builder.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/domain_builder/column_domain_builder.py]: Yes, the return type is `List[Dict[str, Union[str, MetricDomainTypes, Dict[str, Any]]]]` -- removing the comment (in next commit).

File: great_expectations/profiler/domain_builder/column_domain_builder.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/domain_builder/column_domain_builder.py]: I will try to fix this in the next commit.

File: great_expectations/profiler/domain_builder/simple_semantic_type_domain_builder.py (hunk 4 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/domain_builder/simple_semantic_type_domain_builder.py]: 🙇 Changing the code to reflect this clarification.  Thanks!

File: great_expectations/profiler/parameter_builder/multi_batch_parameter_builder.py (hunk 5 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/parameter_builder/multi_batch_parameter_builder.py:32]: @talagluck This is because there was a redundant `exceptions.py` module defined in the `profiler` project; I moved all exceptions to the standard Great Expectations exceptions module.

File: great_expectations/profiler/parameter_builder/parameter.py (hunk 6 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/parameter_builder/parameter.py:8]: 👍 Just added a comment to make the refactor to use `DataClass` in develop -- hope it works!

File: great_expectations/profiler/parameter_builder/parameter.py (hunk 7 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/parameter_builder/parameter.py]: @talagluck Agreed -- a comment to this effect, worded almost identically to yours, is now in this module!

File: great_expectations/profiler/profiler_rule/rule_state.py (hunk 9 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/profiler_rule/rule_state.py:15]: @talagluck I will try this in the next commit.

File: great_expectations/profiler/profiler_rule/rule_state.py (hunk 10 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/profiler_rule/rule_state.py]: @talagluck This is a great question.  We see that `ParameterBuilder` takes `parameter_id` as a parameter in its constructor.  Then we see that `ProfilerRule.evaluate()` uses it as follows: `rule_state.parameters[domain_id][parameter_id] = parameter.parameters`.  However, I have not yet seen `ParameterBuilder` (or any of its descendants) being instantiated.  According to `https://great-expectations--2696.org.readthedocs.build/en/2696/guides/how_to_guides/creating_and_editing_expectations/how_to_create_an_expectation_suite_with_the_user_configurable_profiler.html`, all parameter builder configurations have `id` and `class_name` in common, but then use specific attributes pertaining the logic of each particular parameter builder.  Hence, I will change the constructor argument to `id`, but otherwise will wait for further discussion to agree on the next steps.  Thank you.  /cc @anthonyburdi @gilpasternak35

File: tests/profiler/test_rule_state.py (hunk 11 above)
Comment (alexsherstinsky) [reply tests/profiler/test_rule_state.py:25]: @talagluck In my understanding, this `f45...` string is the `domain ID`.  Thus far, the `domain` is represented by a dictionary.  Maybe it makes sense to have a typed object for it?  Let us discuss this first, before jumping into implementation.  (We have the `DomainBuilder` object.)  Our `RuleState` object has the notions of all `domains` (as a list) and the `active_domain`; for the `active_domain`, `RuleState` provides the `active_domain_id` property (using `IDDict` to generate it, guaranteed to be unique).  The `parameters` property in `RuleState` is a dictionary, whose key is the domain ID with the `Parameter` object being the value.  The test is constructed according to this structure.

File: great_expectations/profiler/profiler_rule/profiler_rule.py (hunk 8 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/profiler_rule/profiler_rule.py]: @talagluck I am seeing that `variables` is just a single `Parameter` object.  We should discuss to determine whether or not this is what we want.

File: great_expectations/profiler/configuration_builder/parameter_id_configuration_builder.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/configuration_builder/parameter_id_configuration_builder.py]: Done.

File: great_expectations/profiler/domain_builder/column_domain_builder.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/domain_builder/column_domain_builder.py]: Not done yet.

File: great_expectations/profiler/profiler_rule/rule_state.py (hunk 9 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/profiler_rule/rule_state.py:15]: Not done yet.

File: great_expectations/profiler/profiler_rule/rule_state.py (hunk 10 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/profiler_rule/rule_state.py]: I made a change that proposes that `parameter_id` is not meaningful and that `parameter_builder` should use `parameter_name` instead of `id`.

File: tests/profiler/test_rule_state.py (hunk 11 above)
Comment (talagluck) [reply tests/profiler/test_rule_state.py:25]: Writing this out to clarify:

- We currently have one domain builder per rule (which can encompass many domains, like all numeric columns, or all columns ending in `_id`.
- RuleState encompasses state for all rules.
- What does the active_domain in RuleState do?
- If I have a dataset with 3 columns - `a`, `b`, and `c`- all of which are numeric. I create a rule that says for each numeric column, I would like to get the min and the max value. We now have three mins, and three maxes, for columns `a`, `b`, and `c`. Can these all be stored in the above RuleState?

Comment (alexsherstinsky) [reply tests/profiler/test_rule_state.py:25]: @talagluck -- On bullet 3: the `active_domain` plays a role in returning the value of the parameter -- when the parameter name fits a certain reserved keyword, such as `$domain.domain_kwargs`.  Also, if the parameter name does not refer a variable (for variables, the parameter name must start with `$variables.`), then the starting (at the top level of the parameter naming hierarchy -- i.e., the entire fully-qualified parameter name string) is determined by using the ID of the active domain as the key to the `parameters` dictionary (domain_id to Parameter object dict), passed to the constructor.  All this suggests to me that the term `Parameter` may not be the best fit for the semantics of this object.  Happy to discuss.

On bullet 2: I could be wrong, but it seems that the answer is affirmative if we consider that the starting parameter name is a fully-qualified parameter name, meaning the dot punctuated set of syllables.  So it results in a hierarchy of keywords and the `Parameter` references, the tail one being the very metric, `min` or `max`, but the entire object trees are distinct.  That is assuming that prior to using `RuleState`, we have created the `parameters` correctly, which we would pass to `RuleState` upon instantiation.  Would be happy to discuss to work through details.

File: great_expectations/profiler/parameter_builder/metric_parameter_builder.py [hunk 12]
```diff
@@ -34,27 +47,35 @@ def _build_parameters(
         :return: a dictionary of format {'parameters': A given resolved metric}
         """
         # Obtaining any necessary domain kwargs from rule state, otherwise using instance var
-        if self._metric_domain_kwargs.startswith("$"):
-            metric_domain_kwargs = rule_state.get_value(self._metric_domain_kwargs)
+        if isinstance(
```
Comment (anthonyburdi) [inline great_expectations/profiler/parameter_builder/metric_parameter_builder.py:50]: ❤️ Love this defensive check

File: great_expectations/profiler/parameter_builder/multi_batch_bootstrapped_metric_distribution_parameter_builder.py [hunk 13]
```diff
@@ -1,60 +1,77 @@
...
+        *,
+        rule_state: Optional[RuleState] = None,
+        validator: Optional[Validator] = None,
+        batch_ids: Optional[List[str]] = None,
+        **kwargs
+    ) -> Parameter:
         samples = []
         for batch_id in batch_ids:
-            metric_domain_kwargs = copy(rule_state.active_domain.domain_kwargs)
+            metric_domain_kwargs = copy(rule_state.active_domain["domain_kwargs"])
```
Comment (anthonyburdi) [inline great_expectations/profiler/parameter_builder/multi_batch_bootstrapped_metric_distribution_parameter_builder.py:68]: Just a general question and not critical to answer before merging this PR: Could we use DictDot for domain_kwargs so that we can access using the dot notation? Would this help us at all (in this case and any others with dicts) in terms of easier refactor later if we change the name over using dict["string"] accessors?

File: great_expectations/profiler/parameter_builder/multi_batch_bootstrapped_metric_distribution_parameter_builder.py [hunk 14]
```diff
@@ -1,60 +1,77 @@
...
             metric_domain_kwargs.update({"batch_id": batch_id})
 
             if self._metric_value_kwargs.startswith("$"):
-                metric_value_kwargs = rule_state.get_value(self._metric_value_kwargs)
+                metric_value_kwargs = rule_state.get_parameter_value(
```
Comment (anthonyburdi) [inline great_expectations/profiler/parameter_builder/multi_batch_bootstrapped_metric_distribution_parameter_builder.py:72]: I just love stuff like this where we are more explicit about the method name.

File: great_expectations/profiler/parameter_builder/multi_batch_bootstrapped_metric_distribution_parameter_builder.py [hunk 15]
```diff
@@ -66,4 +83,7 @@ def _build_parameters(self, *, rule_state, validator, batch_ids, **kwargs):
                 )
             )
 
-        return {"parameters": samples[0]}
+        return Parameter(
+            parameters=samples[0],
```
Comment (anthonyburdi) [inline great_expectations/profiler/parameter_builder/multi_batch_bootstrapped_metric_distribution_parameter_builder.py:88]: Just a suggestion to add a TODO for later
```suggestion
            # TODO: Using the first sample for now, but this should be extended for handling multiple batches
            parameters=samples[0],
```

**Review by anthonyburdi — APPROVED** [review]
This is a great step forward @alexsherstinsky! If it's ok with you, let's merge this and build upon it!

File: great_expectations/profiler/parameter_builder/metric_parameter_builder.py (hunk 12 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/parameter_builder/metric_parameter_builder.py:50]: I think that the credit goes to @jcampbell :-)

File: great_expectations/profiler/parameter_builder/multi_batch_bootstrapped_metric_distribution_parameter_builder.py (hunk 13 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/parameter_builder/multi_batch_bootstrapped_metric_distribution_parameter_builder.py:68]: I like the idea; I think this should be approached in a "wholesale" manner as single-purpose-focused PR.  Happy to discuss.

File: great_expectations/profiler/configuration_builder/parameter_id_configuration_builder.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/configuration_builder/parameter_id_configuration_builder.py]: @talagluck Do you agree with the change made?

File: great_expectations/profiler/domain_builder/column_domain_builder.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/domain_builder/column_domain_builder.py]: Now done!

File: great_expectations/profiler/profiler_rule/rule_state.py (hunk 9 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/profiler_rule/rule_state.py:15]: Actually, @talagluck and @anthonyburdi -- we should discuss this when we get a chance -- to make sure that we choose the best alternative for ourselves.  (Mentioning @gilpasternak35 -- as an FYI -- but no actions requested, since you have midterms now.). Thanks!

File: great_expectations/profiler/parameter_builder/metric_parameter_builder.py (hunk 12 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/parameter_builder/metric_parameter_builder.py:50]: Oh wait...  No, that one was me! :)

File: great_expectations/profiler/profiler_rule/profiler_rule.py (hunk 8 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/profiler_rule/profiler_rule.py]: This is improved upon in PR #2718 /cc @talagluck @anthonyburdi

File: great_expectations/profiler/profiler_rule/rule_state.py (hunk 9 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/profiler_rule/rule_state.py:15]: This is improved upon in PR #2718 /cc @talagluck @anthonyburdi

File: great_expectations/profiler/parameter_builder/multi_batch_bootstrapped_metric_distribution_parameter_builder.py (hunk 13 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/parameter_builder/multi_batch_bootstrapped_metric_distribution_parameter_builder.py:68]: This is now `https://superconductive.atlassian.net/browse/GE-16`
